snuba/datasets/discover.py

```diff
@@ -0,0 +1,247 @@
+from abc import ABC
+from datetime import timedelta
+from typing import Mapping, Sequence
+
+from snuba.clickhouse.columns import (
+    Array,
+    ColumnSet,
+    DateTime,
+    FixedString,
+    Nested,
+    Nullable,
+    String,
+    UInt,
+    UUID,
+)
+from snuba.datasets.dataset import TimeSeriesDataset
+from snuba.datasets.dataset_schemas import DatasetSchemas
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.schemas import Schema
+from snuba.query.extensions import QueryExtension
+from snuba.query.parsing import ParsingContext
+from snuba.query.project_extension import ProjectExtension, ProjectWithGroupsProcessor
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.query.timeseries import TimeSeriesExtension
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
```

Comment:
I think the "having" clause is relevant as well.

Reply:
Yes, we'll be using `having` soon enough in discover2.

Reply:
Tackled in https://github.com/getsentry/snuba/pull/561